### ISSUE TITLE
Configure TemplateResolver's order created in ThymeleafAutoConfiguration

### ISSUE THREAD

**Issue by yacota** [spring-projects/spring-boot#3575]
If order is not setted in ThymeleafAutoConfiguration's method defaultTemplateResolver() then the sort algorithm that Thymelaf uses, in org.thymeleaf.Configuration.TemplateResolverComparator, will leave spring boot's TemplateResolver always before mine, no matter what order I set to my custom TempalteResolver.

**Comment by snicoll** [comment]
Can you share a project that demonstrates the issue please? I am looking at the code and it does not match what you are describing. In particular, I expect `Thymeleaf` to use any `TemplateResolver` that has an order set over one that doesn't (and we don't see any order in Spring Boot).

**Comment by yacota** [comment]
I will, but let me explain the problem in more details.
Spring boot does not set order in the TemplateResolver created in ThymeleafAutoConfiguration class.

    @Bean
		public ITemplateResolver defaultTemplateResolver() {
			TemplateResolver resolver = new TemplateResolver();
			resolver.setResourceResolver(thymeleafResourceResolver());
			resolver.setPrefix(this.properties.getPrefix());
			resolver.setSuffix(this.properties.getSuffix());
			resolver.setTemplateMode(this.properties.getMode());
			resolver.setCharacterEncoding(this.properties.getEncoding());
			resolver.setCacheable(this.properties.isCache());
			return resolver;
		}


Then Thymeleaf order the list of TemplateResolvers received by using org.thymeleaf.Configuration.initialize() method.


            Collections.sort(templateResolversList, TEMPLATE_RESOLVER_COMPARATOR);
            this.templateResolvers = new LinkedHashSet<ITemplateResolver>(templateResolversList);

If you take a look at Thymeleaf's TemplateResolverComparator class you'll see that "null" ordered TemplateResolver's can not be ordered 

     public int compare(final ITemplateResolver o1, final ITemplateResolver o2) {
            // NOTE how they DO NOT CARE about c2 when c1 order is null
            if (o1.getOrder() == null) {
                return -1;
            }
            if (o2.getOrder() == null) {
                return 1;
            }
            return o1.getOrder().compareTo(o2.getOrder());
        }


This means that if I create a custom TemplateResovler it will ALWAYS be used after spring boot's one,  no matter what order I set to mine,no matter if I set null

To sum up I think that ThymeleafAutoConfiguration.defaultTemplateResolver() should set a 0 order

**Comment by snicoll** [comment]
Yeah, that compare implementation looks weird to me. We can certainly do something about it (document the  default order is a good thing) but the (thymeleaf) documentation does not match your description so there is a bug in thymeleaf which is affecting us.

(BTW: the order starts at 1 according to the doc). I'll have a deeper look tomorrow. Thanks!

**Comment by yacota** [comment]
And here is the sample project https://github.com/yacota/spring-boot-thymeleaf

You're welcome

**Comment by snicoll** [comment]
Sorry to be a party pooper but your sample project didn't help much. Lots of unrelated stuff there and copy/pasting our auto-config does not help to figure out what you want to demonstrate.

I built a sample of my own https://github.com/snicoll-scratches/thymeleaf-view-resolver-order

The javadoc of `setOrder` states that:
> Sets a new order for the template engine in the chain. Order should start with 1.

And yet keeps the default value at `null` (which means "first no matter what"). I think we want to keep Thymeleaf's default but I'll offer an extra property to customize the order.

**Comment by snicoll** [comment]
See [this commit](https://github.com/snicoll-scratches/thymeleaf-view-resolver-order/commit/0b3cb7cd22772b4bc2ffbe223b49fe4bd64b6d79) for a sample of the new property.

**Comment by yacota** [comment]
That's all that I want, thanks
